airbyte-integrations/connectors/source-zuora/source_zuora/source.py

```diff
@@ -0,0 +1,166 @@
+        """
+        schema = {}
+        schema["properties"] = self.api._zuora_object_to_json_schema(self.name)
+        return schema
+
+    def as_airbyte_stream(self) -> AirbyteStream:
+        """
+        Override this method to set the default_cursor_field to the stream schema.
+        """
+        stream = super().as_airbyte_stream()
+        stream.default_cursor_field = [self.get_cursor_from_schema(stream.json_schema["properties"])]
+        return stream
+
+    def _get_stream_state(self, stream_state: Mapping[str, Any] = None) -> Mapping[str, Any]:
+        """
+        Get the state of the stream for the default cursor_field = updateddate,
+        If the stream doesn't have 'updateddate', then we use 'createddate'.
+        If stream state == None, than we use start_date by default.
+        """
+        return stream_state.get(self.cursor_field, stream_state.get(self.alt_cursor_field)) if stream_state else self.api.start_date
+
+    def read_records(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping]:
+        # if stream_state is missing, we will use the start-date from config for a full refresh
+        stream_state = self._get_stream_state(stream_state)
+        try:
+            yield from self.api._get_data_with_date_slice("select", self.name, self.cursor_field, stream_state, self.window_in_days)
+        except ZOQLQueryFieldCannotResolve:
+            self.cursor_field = self.alt_cursor_field
+            yield from self.api._get_data_with_date_slice("select", self.name, self.cursor_field, stream_state, self.window_in_days)
+        except ZOQLQueryCannotProcessObject:
+            pass
+
+
+# Incremental-refresh for Zuora Stream handels both full-refresh and incremental-refresh
+class IncrementalZuoraStream(ZuoraStream):
+
+    # Define default cursor field
+    cursor_field = "updateddate"
+    alt_cursor_field = "createddate"
+
+    # setting limit of the date-slice for the data query job
+    @property
+    def window_in_days(self) -> int:
+        return self.api.window_in_days
+
+    # setting checkpoint interval to the limit of date-slice
+    state_checkpoint_interval = window_in_days
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        return {self.cursor_field: max(latest_record.get(self.cursor_field, ""), current_stream_state.get(self.cursor_field, ""))}
+
+
+# Basic Connections Check
+class SourceZuora(AbstractSource):
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, any]:
+        """
+        Testing connection availability for the connector by granting the token.
+        """
+        auth = ZuoraAuthenticator(config["is_sandbox"]).generateToken(config["client_id"], config["client_secret"])
+        if auth.get("status") == 200:
+            return True, None
+        else:
+            return False, auth["status"]
```

Comment:
Is there any message in the response if the connection is failed? 
Could we return this particular message, but not just status code ?

Reply:
If the connection is `failed` due to client bad credentials or out of date credentials (auth credentials were invoked or removed), the actual message is present as the part of the `status`  message and looks like this:

`{"type": "LOG", "log": {"level": "ERROR", "message": "Check failed"}}`
`{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "HTTPError('401 Client Error: Unauthorized for url: https://rest.apisandbox.zuora.com/oauth/token')"}}`

Reply:
Looks good for me